src/System.Management.Automation/engine/hostifaces/PSTask.cs

```diff
@@ -842,6 +876,61 @@ private void CheckForComplete()
             }
         }
 
+        private Runspace GetRunspace(int taskId)
+        {
+            var runspaceName = string.Format(CultureInfo.InvariantCulture, "{0}:{1}", PSTask.RunspaceName, taskId);
+
+            if (_runspacePool.TryDequeue(out Runspace runspace))
+            {
+                if (runspace.RunspaceStateInfo.State == RunspaceState.Opened &&
+                    runspace.RunspaceAvailability == RunspaceAvailability.Available)
+                {
+                    runspace.Name = runspaceName;
+                    return runspace;
+                }
+                RemoveActiveRunspace(runspace);
+            }
+
+            // Create and initialize a new Runspace
+            var iss = InitialSessionState.CreateDefault2();
+            iss.LanguageMode = (SystemPolicy.GetSystemLockdownPolicy() == SystemEnforcementMode.Enforce)
+                ? PSLanguageMode.ConstrainedLanguage : PSLanguageMode.FullLanguage;
+            runspace = RunspaceFactory.CreateRunspace(iss);
+            runspace.Name = runspaceName;
+            runspace.Open();
+            _activeRunspaces.TryAdd(runspace.Id, runspace);
+
+            return runspace;
+        }
+
+        private void ReturnRunspace(PSTaskBase task)
+        {
+            try
+            {
+                var runspace = task.Runspace;
+                Dbg.Assert(runspace != null, "Task runspace cannot be null.");
+                if (runspace.RunspaceStateInfo.State == RunspaceState.Opened &&
+                    runspace.RunspaceAvailability == RunspaceAvailability.Available)
+                {
+                    runspace.ResetRunspaceState();
```

Comment:
How expensive is this?  Perhaps it should be deferred until we need to reuse the runspace?

Reply:
It is pretty fast, but of course that is relative.  It is certainly much faster than creating and initializing a new runspace.  But I don't see any harm in moving it to when used.

Reply:
Oh, now I remember why.  The runspace reset also affects the execution context so needs to be done after use, not before use.